### PR TITLE
First implementation of a Variation Autoencoder CV

### PR DESCRIPTION
**Pull request by andrrizzi**
## Description
This implements a CV based on a standard variational autoencoder using a MSE reconstruction loss and regularizing with a KL divergence w.r.t. a ``Normal(0, 1)`` distribution. For the CV, only the mean is used as CV in production.

## Questions
- [ ] I forgot, did we decide to have a mechanism to do the preprocessing inside ``training_step()``? For now, I'm not doing any preprocessing during training.

## Status
- [x] Ready to go

### CODE REVIEW

File: mlcvs/cvs/unsupervised/vae.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            options['weights'] = train_batch['weights']
+
+        # TODO: Should we do preprocessing here?
+
+        # Encode/decode.
+        mean, log_variance, x_hat = self.encode_decode(x)
+
+        # Loss function.
+        diff = x - x_hat
+        loss = self.loss_function(diff, mean, log_variance, **options)
```
Comment (luigibonati) [inline mlcvs/cvs/unsupervised/vae.py:194]: do you think it makes sense to make the loss function return also the contributions to the loss (reconstruction and KL) and log them?

**Comment by luigibonati** [comment]
> ## Questions
> * [ ]  I forgot, did we decide to have a mechanism to do the preprocessing inside `training_step()`? For now, I'm not doing any preprocessing during training.

No, the way you implemented is correct and uniform with the other classes (e.g. overloading forward_blocks) in the sense that base cv class has a forward method that adds pre and post processing modules to forward_blocks. As it is now, this is not supposed to take place during training

File: mlcvs/cvs/unsupervised/vae.py (hunk 1 above)
Comment (andrrizzi) [reply mlcvs/cvs/unsupervised/vae.py:194]: Yeah, I think that's a good idea. Maybe we could return only the loss as default and control whether the decomposition in kl and reconstruction is returned with an optional argument of ``elbo_gaussians_loss()``?
